web/sw.js

```diff
@@ -0,0 +1,95 @@
+// There can be multiple clients (pages) receiving files, so they generate an id
+// and here we store info assosiated with each transfer.
+const streams = new Map();
+
+// Repurposing the map...
+const waitForMetadata = id => new Promise((resolve, reject) => {
+	streams.set(id, {resolve, reject});
+});
+const signalMetadataReady = (id, streamInfo) => {
+	if (streams.has(id)) streams.get(id).resolve(streamInfo);
+}
+
+const createStream = (onCancel) => {
+	const streamInfo = {};
+
+	streamInfo.stream = new ReadableStream({
+		start (controller) {
+			streamInfo.controller = controller;
+		},
+		cancel (reason) {
+			onCancel(reason);
+		}
+	});
+
+	return streamInfo;
+};
+
+self.addEventListener('message', event => {
+	const message = event.data;
+	const id = message.id;
+
+	if (message.type === 'metadata') {
+		const {name, size} = message;
+
+		// TODO propagate cancellation back to main window and sender.
+		const onCancel = cancelReason => console.warn('Stream cancelled', cancelReason);
+		const streamInfo = { name, size, ...createStream(onCancel) };
+
+		// Resolve promise if GET request arrived first.
+		signalMetadataReady(id, streamInfo);
+
+		streams.set(id, streamInfo);
+
+	} else {
+		const streamInfo = streams.get(id);
+
+		if (message.type === 'data') {
+			streamInfo.controller.enqueue(new Uint8Array(message.data));
+
+		} else if (message.type === 'end') {
+			streamInfo.controller.close();
+
+			// Synchronize with fetch handler to clean up properly.
+			if (streamInfo.requestHandled)
+				streams.delete(id);
+			else streamInfo.streamHandled = true;
+
+		} else if (message.type === 'error') {
+			streamInfo.controller.error(message.error);
+		}
+	}
+});
+
+const encodeFilename = filename =>
+	encodeURIComponent(filename).replace(/['()]/g, escape).replace(/\*/g, '%2A');
```

Comment:
`replace(/\*/g, '%2A')` doesn't need to be a regexp.

i'd also rather we kept it stupid and spelled it out:
```
encodeURIComponent(filename)
  .replace('\'', '%27')
  .replace('(','%28')
  .replace('(','%29')
  .replace('*','%2A')
```
than use `escape` or regexes. (even though it looks kinda silly)

Reply:
Without a regexp with global flag, only first occurrence is replaced. `replaceAll` hasn't shipped on Chrome yet.

Reply:
ah! good point. let's keep it as it then.